src/herder/TransactionQueue.cpp

```diff
@@ -278,6 +278,20 @@ TransactionQueue::canAdd(TransactionFrameBasePtr tx,
                 return TransactionQueue::AddResult::ADD_STATUS_ERROR;
             }
 
+            // Before rejecting Soroban transactions due to source account
+            // limit, check validity of its declared resources, and return an
+            // appropriate error message
+            if (tx->isSoroban())
```

Comment:
Why do we need to check this first? It seems like the root issue we're trying to solve here is when the only thing keeping a Soroban TX out of the queue is invalid resources, and we return `TRY_AGAIN_LATER`. If there are multiple issues with the TX, such as invalid resources and and multiple TXs from the same account, I think it's reasonable to return `TRY_AGAIN_LATER`, since that's a valid actionable error for one of the issues with the TX.

Not a huge deal either way, but moving this check to the end of the function might buy you a few less calls to `checkValid`, but probably not much.

Reply:
mm, I think in the scenario you describe it actually makes more sense to return resource error rather than try again later. in this case, the transaction will never be valid and re-submitting later won't help, so we should probably give the user at least a rough idea about what's wrong.

Reply:
Yeah I agree that this is a more actionable error to the end user, but if you're concerned about pushing too many diagnostic events you could do this check last instead of first and still be correct, which a few less diagnostic events. Probably not worth doing though since it would only save us in a few edge cases.

Reply:
We always attach just a single fixed-size diagnostic event during validation, so I don't think we should be concerned about attaching it. IIUC the main reason for starting the validation from tx queue and not `checkValid` is that signature verification is relatively expensive. Everything else should be much cheaper in comparison.

Reply:
yeah thanks for confirming, that's what @sisuresh concluded as well, so I think we're good to go with this change.